### PR TITLE
Implement avl_union

### PR DESCRIPTION
**Pull request by neithernut**
This PR adds the `avl_union` function and it's implementation.

### CODE REVIEW

File: src/libreset/avl/avl_union.c [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ */
+
+/**
+ * Determine the union of two AVL subtrees (recursively)
+ *
+ * @returns 0 on success, a negative value on error
+ */
+int
+merge_trees(
```
Comment (TheNeikos) [inline src/libreset/avl/avl_union.c:23]: Does merge tree actually change them? (One argument is called dest) Do they get merged in the end or just checked?

File: src/libreset/avl/avl_union.c [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+ * internal implementations
+ *
+ *
+ */
+
+int
+merge_trees(
+    struct avl_el** dest, //!< source subtree
+    struct avl_el const* src, //!< destination subtree
+    struct r_set_cfg const* cfg //!< type information proveded by the user
```
Comment (TheNeikos) [inline src/libreset/avl/avl_union.c]: `proveded` should be `provided`

File: src/libreset/avl/avl_union.c [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+) {
+    // if the source is empty, all it's items are integrated into dest
```
Comment (TheNeikos) [inline src/libreset/avl/avl_union.c:64]: `it's` should be `its`

File: src/libreset/avl/avl_union.c (hunk 1 above)
Comment (neithernut) [inline src/libreset/avl/avl_union.c:23]: Elements which are not in `dest` are inserted into `dest`.
`src` is not modified.

File: src/libreset/avl/avl.h [hunk 4]
```diff
@@ -147,6 +147,31 @@ __r_nonnull__(1)
...
+ * Elements which are already in the destination AVL will not be copied.
+ * This is equivalent to determining the union of both AVLs and storing the
+ * result in the destination.
+ * The function will not modify the source AVL.
+ *
+ * @note This function may also be used to copy an AVL into another
+ *
+ * @memberof avl
+ *
+ * @return 0 if the operation was successful, some negative value on error
```
Comment (matthiasbeyer) [inline src/libreset/avl/avl.h]: I would like to see a list what error numbers on what kind of error, as I do in #247

File: src/libreset/avl/avl_union.c [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * forward declarations
+ *
+ *
+ */
+
+/**
+ * Determine the union of two AVL subtrees (recursively)
+ *
+ * @returns 0 on success, a negative value on error
```
Comment (matthiasbeyer) [inline src/libreset/avl/avl_union.c]: Also here.

File: src/libreset/avl/avl_union.c (hunk 3 above)
Comment (neithernut) [inline src/libreset/avl/avl_union.c:64]: I think it's fine as it is (`it's items` ~ `the source's`)

File: src/libreset/avl/avl_union.c [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ *
+ * internal implementations
+ *
+ *
+ */
+
+int
+merge_trees(
+    struct avl_el** dest, //!< source subtree
```
Comment (matthiasbeyer) [inline src/libreset/avl/avl_union.c]: The actual argument documentation goes into the interface definition, not the implementation!

Comment (matthiasbeyer) [inline src/libreset/avl/avl_union.c]: The documentation is also wrong here. `dest` is the source` what?

File: src/libreset/avl/avl_union.c [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    struct avl_el const* src, //!< destination subtree
```
Comment (matthiasbeyer) [inline src/libreset/avl/avl_union.c]: Same here: `src` is the dest?

File: src/libreset/avl/avl_union.c [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    retval = merge_trees(&(*dest)->r, src->r, cfg);
+    if (retval < 0) {
+        goto cleanup;
+    }
+
+    // we have matching nodes and may now merge the corresponding lists
+    retval = ll_union(&(*dest)->ll, &src->ll, cfg);
+
+    // clean up
+    cleanup:
```
Comment (matthiasbeyer) [inline src/libreset/avl/avl_union.c]: Wouldn't it be better to remove the indentation level here? I think it is visually much nicer, as you can find the label more easily...

**Comment by neithernut** [comment]
Please give me an ACK for squashing as soon as you agree with the changes.

**Comment by matthiasbeyer** [comment]
ACK. Please squash now.

**Comment by neithernut** [comment]
Done.

**Pull request merged**